src/Mod/Sketcher/Gui/ViewProviderSketch.cpp

```diff
@@ -1212,45 +1212,165 @@ bool ViewProviderSketch::mouseButtonPressed(int Button, bool pressed, const SbVe
                     return true;
                 case STATUS_NONE: {
                     // A right click shouldn't change the Edit Mode
-                    if (preselection.isPreselectPointValid()) {
-                        return true;
+                    bool noSelection = true;
+                    bool selectPoint = false;
+                    bool selectLine = false;
+                    bool selectConstraint = false;
+                    bool selectMultiple = false;
+
+                    Gui::MenuItem geom;
+                    geom.setCommand("Sketcher context");
+                    // Gets a selection vector
+                    std::vector<Gui::SelectionObject> selection = Gui::Selection().getSelectionEx();
+                    // if something is selected
+                    if (selection.size() > 0) {
+                        const std::vector<std::string> SubNames = selection[0].getSubNames();
+
+                    if (SubNames.size() == 1) {
+                        noSelection = false;
+                        if (SubNames.begin()->substr(0, 4) == "Edge")
+                            selectLine = true;
+                        else if (SubNames.begin()->substr(0, 4) == "Vert")
+                            selectPoint = true;
+                        else if (SubNames.begin()->substr(0, 4) == "Cons")
+                            selectConstraint = true;
                     }
-                    else if (preselection.isPreselectCurveValid()) {
-                        return true;
+                    else if (SubNames.size() > 1) {
+                        noSelection = false;
+                        selectMultiple = true;
+                        int selectedLines = 0;
+                        int selectedPoints = 0;
+                        int selectedConstraints = 0;
+                        // go through the selected subelements
+                        for (std::vector<std::string>::const_iterator it = SubNames.begin();
+                            it != SubNames.end();
+                            ++it) {
+                            if (it->substr(0, 4) == "Edge") {
+                                ++selectedLines;
+                            }
+                            else if (it->substr(0, 4) == "Vert"){
+                                ++selectedPoints;
+                            }
+                            else if (it->substr(0, 4) == "Cons"){
+                                ++selectedConstraints;
+                            }
+                        }// End of Iteration
+                        if (selectedLines > 1 && selectedPoints == 0 && selectedConstraints == 0)
```

Comment:
I don't see the value of using the boolean selectLine, selectMultiple etc. (besides selectMultiple makes you have to use the special case of `== 1`)
Instead use directly the conditions you are using now in the menu creations
`if (selectedLines > 1 && selectedPoints == 0 && selectedConstraints == 0)`